### DISCUSSION TITLE
In the project, it was accidentally discovered that postgres+flask_sqlalchemy has a memory leak problem when querying

### DISCUSSION THREAD

**Discussion by mokesher** [sqlalchemy/sqlalchemy#8542] [Usage Questions]
### Describe the bug

Let's talk about the environment first: 
1. python3.6-python3.9 are all used, and problems are found. There are no irrecyclable objects printed by gc, which excludes the problem of python itself. 
2. SQLAlchemy version 1.3-latest 1.4.41 has the problem that the memory is not released after querying. In particular, using native sql query will release much less memory than orm model query. The database table stores more than 6000 pieces of data for testing. In the normal operation of the project, the memory will gradually increase for a period of time because the memory of sqlalchemy is not released. Until oom, I have used a lot of memory analysis tools to locate the following. I don't know how to solve it. Large library, there is such a problem has not been found

There are problems with various query statements referring to the official documentation of sqlalchemy. The demo written is to start a multi-thread, and then query. After the thread is destroyed, only a part of the memory is released. After more than 6000 pieces of data orm are checked, there are still 25M that have not been released.

### To Reproduce

```python
# -*- coding: utf-8 -*-
import enum
import gc
import os
import threading
# from sqlalchemy.orm import Session
import time
from contextlib import contextmanager

import psutil
from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import Column, VARCHAR, Text

# metadata = MetaData()
# Base = declarative_base()

user = "postgres"
pwd = ""
host = ""
port = ""
database = ""
db_uri = f"postgresql+psycopg2://{user}:{pwd}@{host}:{port}/{database}"

app = Flask(__name__)
app.config["SQLALCHEMY_DATABASE_URI"] = db_uri
app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
app.config["SQLALCHEMY_RECORD_QUERIES"] = False
# app.config["SQLALCHEMY_ECHO"] = True
engine_options = {
    "pool_size": 50,
    "pool_timeout": 30,
    "pool_recycle": 2 * 60 * 60,
    "max_overflow": 10,
    # "echo": True,
    "query_cache_size": 0,
    "execution_options":
        {
            "compiled_cache": None,
        }
}

app.config["SQLALCHEMY_ENGINE_OPTIONS"] = engine_options
db = SQLAlchemy(app, session_options={"expire_on_commit": False})


# engine = create_engine(db_uri)
# engine = db.get_engine()
# engine = create_engine(db_uri, future=True, **engine_options)
# Session = sessionmaker(autoflush=True, bind=engine)

@enum.unique
class TaskStatus(enum.Enum):
    INIT = "INIT"
    RUNNING = "RUNNING"
    SUCCESS = "SUCCESS"
    FAILED = "FAILED"
    FAILED_NEED_ROLLBACK = "FAILED_NEED_ROLLBACK"
    FAILED_NEED_CLEAR = "FAILED_NEED_CLEAR"
    TIMEOUT = "TIMEOUT"
    PARTIAL_SUCCESS = "PARTIAL_SUCCESS"
    PARTIAL_SUCCESS_NEED_CLEAR = "PARTIAL_SUCCESS_NEED_CLEAR"
    CLEAN_SUCCESS = "CLEAN_SUCCESS"
    CLEAN_FAILED = "CLEAN_FAILED"
    ROLLBACK_SUCCESS = "ROLLBACK_SUCCESS"
    ROLLBACK_FAILED = "ROLLBACK_FAILED"


class TaskModel(db.Model):
    __tablename__ = "task"
    id = Column(VARCHAR(64), primary_key=True)  # uuid
    task_id = Column(VARCHAR(32), nullable=False)  # Internal definition
    name = Column(VARCHAR(128))
    description = Column(Text)

    ***省略


@contextmanager
def get_session():
    try:
        # session = Session()
        session = db.session
        yield session
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()
        session.remove()


def sleep_fun():
    print("sleep_fun")
    global begin_rss
    begin_rss = psutil.Process(os.getpid()).memory_info().rss / 1024 / 1024

    for j in range(1):
        print(j)

        with get_session() as session:
            a = session.execute("select * from task").all()
            # a = session.query(TaskModel).all()
            print(psutil.Process(os.getpid()).memory_info().rss / 1024 / 1024)
            print(2222, len(a))

    print("begin", begin_rss)
    print("end", psutil.Process(os.getpid()).memory_info().rss / 1024 / 1024)
    print("finish!!")


t = threading.Thread(target=sleep_fun)
t.start()
t.join()
gc.collect()

print("end")

end_rss = psutil.Process(os.getpid()).memory_info().rss / 1024 / 1024
print(end_rss)

print("cost: ", end_rss - begin_rss)
time.sleep(1000000)
```

### Error

In particular, using native sql query will release much less memory than orm model query
use orm query 
a = session.query(TaskModel).all() 
result： 
begin 57.20703125 
end 106.94140625 
finish!! 
end 82.31640625 
cost: 25.109375

use sql： 
a = session.execute("select * from task").all() 
begin 57.2109375 
end 104.3515625 
finish!! 
end 61.8046875 
cost: 4.59375

I downloaded the source code of sqlalchemy 1.4.40 before, and the breakpoint found that the memory could not be released in the last loading.instances of context.py,

```
@classmethod
    def orm_setup_cursor_result(
        cls,
        session,
        statement,
        params,
        execution_options,
        bind_arguments,
        result,
    ):
        execution_context = result.context
        compile_state = execution_context.compiled.compile_state

        # cover edge case where ORM entities used in legacy select
        # were passed to session.execute:
        # session.execute(legacy_select([User.id, User.name]))
        # see test_query->test_legacy_tuple_old_select

        load_options = execution_options.get(
            "_sa_orm_load_options", QueryContext.default_load_options
        )
        if compile_state.compile_options._is_star:
            return result

        querycontext = QueryContext(
            compile_state,
            statement,
            params,
            session,
            load_options,
            execution_options,
            bind_arguments,
        )

        return loading.instances(result, querycontext)
```

loading.py
After running to this part of the code, the memory cannot be released.

    try:
        (process, labels, extra) = list(
            zip(
                *[
                    query_entity.row_processor(context, cursor)
                    for query_entity in context.compile_state._entities
                ]
            )
        )
        import sys
        print(process, labels, extra)
        if context.yield_per and (
            context.loaders_require_buffering
            or context.loaders_require_uniquing
        ):
            raise sa_exc.InvalidRequestError(
                "Can't use yield_per with eager loaders that require uniquing "
                "or row buffering, e.g. joinedload() against collections "
                "or subqueryload().  Consider the selectinload() strategy "
                "for better flexibility in loading objects."
            )

### Versions

- OS:centos7
- Python:3.6-3.9
- SQLAlchemy:1.3-latest 1.4.41
- Database:postgres
- DBAPI (eg: psycopg, cx_oracle, mysqlclient):

**Comment by zzzeek** [comment]
hey there - 

I have to apologize, I'm not really sure what I'm looking at.      Seems like an ordinary program, it does not seem to load any data into "task", so I dont have any insight on what "6000 pieces of ORM data" refers towards, and seems to just run a query that as given would fail because the "task" table won't exist.    Seems to use flask-sqlalchemy (which we don't support here) but not sure (`# session = Session()` then `session = db.session` - is that important?)     The pstats data there doesn't indicate any issue to me either, the ORM builds up a lot of internal information surrounding each mapped class like TaskModel the first time it's used, so it's normal that memory use would be a little higher after queries were run for the first time.

So here we can go in two directions.  If we want to just figure out what's happening on your end so you can fix your problem, we'd want to poke at changing things and seeing where things work or don't work - starting with does using flask-sqlalchemy "db." or not change things, how fast is memory going up as a function of how many threads / iterations per meg of RAM used, what driver are you using, etc.

If we want to go with, "there's a bug in SQLAlchemy", which certainly, we have memory issues reported a few times a year sure, though they are all very esoteric issues with very specific triggers (all have been fixed of course),  we would need a succinct test case that runs in a loop (has to run, include whatever data it needs, can't use Flask), where I can then run `top` and watch memory go up without ever stopping.

**Comment by shihai1991** [comment]
How about this example? I combine [mokesher](https://github.com/mokesher)'s code with #4616's code.
```
import os
import time
import gc

import psutil

from sqlalchemy import create_engine
from sqlalchemy import Column, SmallInteger, String
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class Test(Base):
    __tablename__ = u'test'

    id = Column('id', SmallInteger, primary_key=True)
    text = Column('text', String(255), nullable=False)


engine = create_engine('sqlite:////tmp/test.sqlite')

Base.metadata.create_all(engine)

SessionMaker = sessionmaker(bind=engine)


def repopulate_db():
    session = SessionMaker()
    q = session.query(Test)
    q.delete()

    for i in range(0, 100000):
        example_text = str(i * i * i)
        session.add(Test(id=i, text=example_text))

    session.commit()
    session.close()


def test_mem_leak():
    print('start...')
    begin_rss = psutil.Process(os.getpid()).memory_info().rss / 1024 / 1024
    print(f'begin rss: {begin_rss}MB')
    session = SessionMaker()
    result = session.query(Test).all()
    print('flush...')
    session.flush()
    session.expunge_all()
    session.close()
    print('sleep...')
    time.sleep(10)
    gc.collect()
    end_rss = psutil.Process(os.getpid()).memory_info().rss / 1024 / 1024
    print(f'end rss: {end_rss}MB')
    print('end...')

if __name__ == "__main__":
    #repopulate_db()
    test_mem_leak()
```

The result in my vm:
```
start...
begin rss: 34.39453125MB
flush...
sleep...
end rss: 185.625MB
end...
```

Versions
OS: centos7
Python: 3.10
SQLAlchemy: 1.4.41.dev0
Database: sqlite

**Reply by zzzeek** [reply]
OK, that is a much better test case.  It still doesn't run in a loop to show a "leak", so I just added a second run at the end:

```py
if __name__ == "__main__":
    #repopulate_db()  # edit: duh, uncomment this
    test_mem_leak()


    test_mem_leak()
```

However it didnt show any kind of growth as yours did:

```
$ python test3.py 
start...
begin rss: 36.5703125MB
flush...
sleep...
end rss: 37.70703125MB
end...
start...
begin rss: 37.70703125MB
flush...
sleep...
end rss: 37.70703125MB
end...
```

it did grow a little bit, that's mostly SQLite's own overhead.   For comparison, here's on a MySQL db ``mysql://scott:tiger@localhost/test``

```
$ python test3.py 
start...
begin rss: 34.04296875MB
flush...
sleep...
end rss: 34.30078125MB
end...
start...
begin rss: 34.30078125MB
flush...
sleep...
end rss: 34.30078125MB
end...
```

also I would note that the behavior of psutil in these runs seems to be very inconsistent, reporting different numbers on every run and sometimes not changing at all.   memory leaks can only be reliably diagnosed by looking for cumulative, non-stop growth.

so we can see sqlite has it's own overhead, which is possibly because [it is caching disk pages in memory](https://www.sqlite.org/pragma.html#pragma_cache_size) and the database engine itself is in the same memory space as your Python program.

so what we are going to look at here is that you're on centos7 which has notoriously old versions of SQLite.    let's see what this shows:

```py
$ python
Python 3.10.0 (default, Nov  5 2021, 17:23:47) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> sqlite3.sqlite_version
'3.36.0'
```

next, can you run your program with:   A. a different database backend, like MySQL or PostgreSQL, and/or B. on a different machine (like a recent ubuntu or fedora, a container is fine) that has a more modern sqlite version.      I tried running your program on a centos:7 container here but also did not replicate those results.

**Comment by shihai1991** [comment]
> can you run your program with: A. a different database backend, like MySQL or PostgreSQL, and/or B. on a different machine

I run this example in centos 8.2.
version infos:
- OS: centos 8.2
- psutil: 5.9.2
- PyMySQL: 1.0.2
- SQLAlchemy: 1.4.41.dev0
- connection config: mysql+pymysql://root:pwd@ip/db')

And get the output info:
```
start...
begin rss: 35.05859375MB
flush...
sleep...
end rss: 73.640625MB
end...
start...
begin rss: 40.140625MB
flush...
sleep...
end rss: 90.6328125MB
end...
```

the sqlite3 version in my centos8.2 is 2.6.0.
When I use this sqlite3 and get the output info:
```
start...
begin rss: 34.7421875MB
flush...
sleep...
end rss: 147.71875MB
end...
start...
begin rss: 45.46875MB
flush...
sleep...
end rss: 195.91796875MB
end...
```

Weird~

**Reply by zzzeek** [reply]
@CaselIT can you confirm the results I'm getting?

**Reply by zzzeek** [reply]
that's not your SQLite version, that's the pysqlite version.   check your SQLite version like this, make sure to type ``sqlite_version`` and not ``version``:

```py
$ python
Python 3.10.0 (default, Nov  5 2021, 17:23:47) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> sqlite3.sqlite_version
'3.36.0'

```

**Reply by shihai1991** [reply]
> that's not your SQLite version, that's the pysqlite version. check your SQLite version like this, make sure to type `sqlite_version` and not `version`:
> 
> ```python
> $ python
> Python 3.10.0 (default, Nov  5 2021, 17:23:47) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import sqlite3
> >>> sqlite3.sqlite_version
> '3.36.0'
> ```

Oh, Thanks for your notice.
```
(venv) [root@sqlalchemy-test venv]# python
Python 3.9.6 (default, Aug 25 2021, 16:22:38)
[GCC 8.5.0 20210514 (Red Hat 8.5.0-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> sqlite3.sqlite_version
'3.26.0'
```

**Reply by CaselIT** [reply]
I've changed the code as
```py
if __name__ == "__main__":
    repopulate_db()
    for _ in range(25):
        test_mem_leak()
```
The output is below. There does not seem to be any leak. The memory increases a bit in the few runs, but after that it stabilized.
<details>
<summary> python log </summary>

```
MovedIn20Warning: The ``declarative_base()`` function is now available as sqlalchemy.orm.declarative_base(). (deprecated since: 2.0) (Background on SQLAlchemy 2.0 at: https://sqlalche.me/e/b8d9)
  Base = declarative_base()
start...
begin rss: 46.8671875MB
flush...
sleep...
end rss: 164.7109375MB
end...
start...
begin rss: 48.5234375MB
flush...
sleep...
end rss: 164.55078125MB
end...
start...
begin rss: 49.7890625MB
flush...
sleep...
end rss: 164.58203125MB
end...
start...
begin rss: 49.81640625MB
flush...
sleep...
end rss: 164.58203125MB
end...
start...
begin rss: 49.56640625MB
flush...
sleep...
end rss: 164.58203125MB
end...
start...
begin rss: 49.81640625MB
flush...
sleep...
end rss: 164.42578125MB
end...
start...
begin rss: 49.41015625MB
flush...
sleep...
end rss: 164.50390625MB
end...
start...
begin rss: 49.73828125MB
flush...
sleep...
end rss: 164.51171875MB
end...
start...
begin rss: 49.99609375MB
flush...
sleep...
end rss: 164.51953125MB
end...
start...
begin rss: 50.00390625MB
flush...
sleep...
end rss: 164.51953125MB
end...
start...
begin rss: 50.00390625MB
flush...
sleep...
end rss: 164.53125MB
end...
start...
begin rss: 49.765625MB
flush...
sleep...
end rss: 164.53125MB
end...
start...
begin rss: 50.015625MB
flush...
sleep...
end rss: 164.53125MB
end...
start...
begin rss: 49.765625MB
flush...
sleep...
end rss: 164.53125MB
end...
start...
begin rss: 49.765625MB
flush...
sleep...
end rss: 164.53125MB
end...
start...
begin rss: 50.015625MB
flush...
sleep...
end rss: 164.49609375MB
end...
start...
begin rss: 49.98046875MB
flush...
sleep...
end rss: 164.5MB
end...
start...
begin rss: 49.984375MB
flush...
sleep...
end rss: 164.51171875MB
end...
start...
begin rss: 50.24609375MB
flush...
sleep...
end rss: 164.515625MB
end...
start...
begin rss: 50.0MB
flush...
sleep...
end rss: 164.51953125MB
end...
start...
begin rss: 49.75390625MB
flush...
sleep...
end rss: 164.51953125MB
end...
start...
begin rss: 49.7578125MB
flush...
sleep...
end rss: 164.52734375MB
end...
start...
begin rss: 50.26171875MB
flush...
sleep...
end rss: 164.52734375MB
end...
start...
begin rss: 49.76171875MB
flush...
sleep...
end rss: 164.52734375MB
end...
start...
begin rss: 49.51171875MB
flush...
sleep...
end rss: 164.52734375MB
end...
```
</details>

**Reply by shihai1991** [reply]
I use another centos8.4 container and run `test_mem_leak()` 15 times.
Env version info:
- OS: CentOS Linux release 8.4.2105
- Python: 3.9.6
- psutil: 5.9.2
- PyMySQL: 1.0.2
- SQLAlchemy: 1.4.41

Using mysql+pymysql to test, and the output:
```
start...
begin rss: 28.95703125MB
flush...
sleep...
end rss: 67.4375MB
end...
start...
begin rss: 34.4375MB
flush...
sleep...
end rss: 84.4453125MB
end...
start...
begin rss: 57.4453125MB
flush...
sleep...
end rss: 84.46484375MB
end...
start...
begin rss: 57.21484375MB
flush...
sleep...
end rss: 84.47265625MB
end...
start...
begin rss: 57.22265625MB
flush...
sleep...
end rss: 84.48046875MB
end...
start...
begin rss: 57.23046875MB
flush...
sleep...
end rss: 84.484375MB
end...
start...
begin rss: 57.234375MB
flush...
sleep...
end rss: 84.48828125MB
end...
start...
begin rss: 57.23828125MB
flush...
sleep...
end rss: 84.4921875MB
end...
start...
begin rss: 57.2421875MB
flush...
sleep...
end rss: 84.49609375MB
end...
start...
begin rss: 57.24609375MB
flush...
sleep...
end rss: 84.49609375MB
end...
start...
begin rss: 57.24609375MB
flush...
sleep...
end rss: 84.5MB
end...
start...
begin rss: 57.25MB
flush...
sleep...
end rss: 84.5MB
end...
start...
begin rss: 57.25MB
flush...
sleep...
end rss: 84.50390625MB
end...
start...
begin rss: 57.25390625MB
flush...
sleep...
end rss: 84.5078125MB
end...
start...
begin rss: 57.2578125MB
flush...
sleep...
end rss: 84.5078125MB
end...
```

In my another centos8.2 vm:
Env version info: 
- OS: CentOS Linux release 8.2.2004 (Core)
- Python: 3.9.6
- psutil: 5.9.2
- PyMySQL: 1.0.2
- SQLAlchemy: 1.4.41.dev0
Using mysql+pymysql to test, and the output:
```
start...
begin rss: 35.28125MB
flush...
sleep...
end rss: 73.86328125MB
end...
start...
begin rss: 40.36328125MB
flush...
sleep...
end rss: 90.85546875MB
end...
start...
begin rss: 63.609375MB
flush...
sleep...
end rss: 90.875MB
end...
start...
begin rss: 63.875MB
flush...
sleep...
end rss: 90.88671875MB
end...
start...
begin rss: 63.88671875MB
flush...
sleep...
end rss: 90.95703125MB
end...
start...
begin rss: 63.95703125MB
flush...
sleep...
end rss: 90.97265625MB
end...
start...
begin rss: 64.22265625MB
flush...
sleep...
end rss: 90.9765625MB
end...
start...
begin rss: 63.7265625MB
flush...
sleep...
end rss: 90.98046875MB
end...
start...
begin rss: 63.98046875MB
flush...
sleep...
end rss: 90.98046875MB
end...
start...
begin rss: 63.98046875MB
flush...
sleep...
end rss: 90.98046875MB
end...
start...
begin rss: 63.98046875MB
flush...
sleep...
end rss: 90.98046875MB
end...
start...
begin rss: 63.98046875MB
flush...
sleep...
end rss: 90.984375MB
end...
start...
begin rss: 64.484375MB
flush...
sleep...
end rss: 90.984375MB
end...
start...
begin rss: 64.234375MB
flush...
sleep...
end rss: 90.98828125MB
end...
start...
begin rss: 64.48828125MB
flush...
sleep...
end rss: 90.98828125MB
end...
```

In my old centos7(CentOS Linux release 7.2.1511 (Core)) vm, Using mysql+pymysql to test and the output:
```
start...
begin rss: 35.46875MB
flush...
sleep...
end rss: 88.0859375MB
end...
start...
begin rss: 45.76953125MB
flush...
sleep...
end rss: 105.85546875MB
end...
start...
begin rss: 48.984375MB
flush...
sleep...
end rss: 106.94921875MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 107.9140625MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 106.94921875MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 107.9140625MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 106.94921875MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 107.9140625MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 106.94921875MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 107.9140625MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 106.94921875MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 107.9140625MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 106.94921875MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 107.9140625MB
end...
start...
begin rss: 49.09375MB
flush...
sleep...
end rss: 106.94921875MB
end...
```

**Reply by CaselIT** [reply]
The memory usage is constant after the first few runs, so it seems there is no leak here.

**Comment by shihai1991** [comment]
> The memory increases a bit in the few runs, but after that it stabilized.

So the memory increases a bit is a normal condition at first? (excluding the sqlite3 as Mike said in https://github.com/sqlalchemy/sqlalchemy/discussions/8542#discussioncomment-3708972)

**Reply by shihai1991** [reply]
I've updated the main funcion to:
```python
if __name__ == "__main__":
    #repopulate_db()
    threads = []
    for _ in range(15):
        thread = threading.Thread(target=test_mem_leak)
        thread.start()
        threads.append(thread)
    for thread in threads:
        thread.join()

    time.sleep(10)
    gc.collect()
    main_end_rss = psutil.Process(os.getpid()).memory_info().rss / 1024 / 1024
    print(f'the main thread end rss: {main_end_rss}MB')
```

the output result in my centos7, looks like the rss of multithreading's query is more large:
```
start...
start...
start...
start...
start...
start...
start...
start...
begin rss: 35.7421875MB
begin rss: 35.7421875MB
begin rss: 35.7421875MB
begin rss: 35.7421875MB
begin rss: 35.7421875MB
start...
begin rss: 36.07421875MB
begin rss: 35.7421875MB
start...
start...
start...
start...
start...
begin rss: 36.49609375MB
begin rss: 35.99609375MB
begin rss: 36.49609375MB
begin rss: 36.49609375MB
begin rss: 36.49609375MB
begin rss: 36.49609375MB
begin rss: 36.49609375MB
start...
begin rss: 37.04296875MB
flush...
sleep...
end rss: 309.921875MB
end...
flush...
sleep...
flush...
end rss: 391.4140625MB
end...
sleep...
flush...
sleep...
end rss: 445.8125MB
end...
end rss: 468.62109375MB
end...
flush...
sleep...
flush...
flush...
flush...
end rss: 566.83203125MB
end...
sleep...
sleep...
sleep...
end rss: 579.65625MB
end...
end rss: 576.8046875MB
end...
end rss: 571.671875MB
end...
flush...
flush...
sleep...
sleep...
flush...
flush...
sleep...
sleep...
end rss: 569.890625MB
end...
end rss: 566.8359375MB
end...
end rss: 554.0390625MB
flush...
end...
flush...
end rss: 532.83984375MB
end...
sleep...
sleep...
flush...
sleep...
end rss: 502.359375MB
end...
end rss: 502.359375MB
end...
end rss: 457.578125MB
end...
the main thread end rss: 86.6875MB
```

When I update the threads num, the progress' overhead is different:
- When the threads num is 1, the main thread's rss: 42.61328125MB
- When the threads num is 5, the main thread's rss: 53.25MB
- When the threads num is 15, the main thread's rss: 86.6875MB

**Reply by CaselIT** [reply]
The connections are pooled, so that may account for the difference. Also the thread list is not cleared so that explains the rest of the memory 
I'm not sure what you are trying to showcase here. The examples you are using cannot determine if there is a leak, they just show that memory is used after an operation.
Could you clarify the question?

**Reply by shihai1991** [reply]
> Could you clarify the question?

I am just curious about why the memory woud be increased when the threads' num increased.

> The connections are pooled, so that may account for the difference.

Sorry, I am unfamiliar with sqlalchemy. The connections are pooled. So some resources of connections in multithreads would be pooled and don't released when thread exit?

> Also the thread list is not cleared so that explains the rest of the memory

After `thread.join()`, the thread list would be cleared.

**Reply by CaselIT** [reply]
> So some resources of connections in multithreads would be pooled and don't released when thread exit?

That's unrelated to threading. If you open multiple connection some of them will be kept open in the pool, since a queue pool is used when connecting to most databases by sqlalchemy.

> After thread.join(), the thread list would be cleared.

In your code `threads` is not cleared so all thread instances are still in the list.

**Reply by shihai1991** [reply]
> That's unrelated to threading. If you open multiple connection some of them will be kept open in the pool, since a queue pool is used when connecting to most databases by sqlalchemy.

OK, got it. Thanks for your explain.

> In your code `threads` is not cleared so all thread instances are still in the list.

Oh, thanks. I forget del the `threads` vars finally.

**Comment by shihai1991** [comment]
I don't have more questions. Thanks Mike, Federico for your quick response :)

**Reply by zzzeek** [reply]
it is still very strange how you have different results on all your platforms, are you sure the script you're testing with is exactly identical to the one you gave us here?  I just ran this on a totally different physical machine with python 3.9 as you are running.  what is your base OS ?   are you running containers under OSX or something like that?

edit: seems like uncommenting "write the data" might help !

**Reply by CaselIT** [reply]
I think the difference compared to your run may be because in the script the function that populates the data was commented out. If you uncomment it you should get similar numbers

**Reply by zzzeek** [reply]
doh, wow, indeed.

**Reply by shihai1991** [reply]
> it is still very strange how you have different results on all your platforms, are you sure the script you're testing with is exactly identical to the one you gave us here? I just ran this on a totally different physical machine with python 3.9 as you are running. what is your base OS ? are you running containers under OSX or something like that?
> 
> edit: seems like uncommenting "write the data" might help !

> what is your base OS ? are you running containers under OSX or something like that?

I use centos7, centos8.2, centos8.4 in docker(the base OS is centos8.2).
I use python3.10 of developing version and disable `pymalloc`(to disable the arenas) in centos7.
In fact, Forbiding the [`pymalloc`](https://docs.python.org/3/c-api/memory.html#pymalloc), the memory cost of process would be increased.

> edit: seems like uncommenting "write the data" might help !

OK, I haven't focus on this detail. In fact, I comment the "write the date" when I use mysql.
